### ISSUE TITLE
InsertOrReplace index problem

### ISSUE THREAD

**Issue by arthurhsu** [google/lovefield#175]
Originally reported in https://groups.google.com/forum/#!topic/lovefield-users/1XHYmiTxZxE
Minimal repro:

https://github.com/peterbe/react-buggy
type mozilla, crontabber
click first issue
refresh, click first issue, Exception 109 is thrown

Faulting is at lf.index.BTreeNode_.prototype.insert. Same key/value is attempted to be inserted into the BTree again causing error 109.

**Comment by freshp86** [comment]
@peterbe: I have been investigating this issue. I am able to reproduce. Not  100% sure yet, but it seems that the problem could be that you are modifying/re-using directly results from previous SELECT queries. For example [here](https://github.com/peterbe/react-buggy/blob/master/src/App.js#L213) (yes, bad idea), but I think this is also happening for the "Issue" table which is causing the error later.

This goes against the spec, see https://github.com/google/lovefield/blob/master/docs/spec/04_query.md#41-select-query-builder, specifically the sentence

"One important concept is to treat the returned results of select queries as read-only and do not modify it. Based on performance considerations, Lovefield does not actively clone/freeze the object retrieved from its internal cache. For this to work, the user is supposed to follow the rule of not altering results returned from select query"

In short, it seems that after such an illegal modification, the in-memory cache is inconsistent with the in-memory indices which leads to the assertion error. Could you eliminate such modifications/reuse first? Then I can keep looking if the problem still happens.

**Comment by freshp86** [comment]
Found the exact culprit, which verifies my previous suspicion. See lines https://github.com/peterbe/react-buggy/blob/master/src/App.js#L329 and L340, where the Date object that was returned from a SELECT query is modified (increased by one second). This corrupts the cache for the issue that has the MAX updated_at value and it no longer matches what is stored in IDB or in the in-memory indices.

Simple fix is to not modify the returned Date object, but to create a new one instead. Here is the fix

```
diff --git a/src/App.js b/src/App.js
index 3f6ab27..251ed5b 100644
--- a/src/App.js
+++ b/src/App.js
@@ -335,12 +335,13 @@ export default class App extends Component {
         }
       }
       if (since) {
+        let newSince = new Date(since);
         // GitHub uses the since= as >= meaning that it will always return
         // something. Increment it by 1 second.
-        since.setSeconds(since.getSeconds() + 1);
-        since = since.toISOString()
+        newSince.setSeconds(newSince.getSeconds() + 1);
+        newSince = newSince.toISOString()
         url += url.indexOf('?') > -1 ? '&' : '?'
-        url += `since=${since}`
+        url += `since=${newSince}`
       }
       console.log("Downloading issues from:", url);
       return downloadIssues(url).then(() => {
```

**Comment by peterbe** [comment]
That's wonderful news! Thank you. 
I'm sorry this was a distraction. Fortunate in my case I had put a in a `// mutating the result. Bad idea??` comment for my own hacking notes. But this could re-appear when people write code. 

Perhaps we can take this as a tip to improve the error message or something. I'm sure I'm not the last one to make that kinda mistake in implementation :)

Great work guys!

**Comment by peterbe** [comment]
That's wonderful news! Thank you. 
I'm sorry this was a distraction. Fortunate in my case I had put a in a `// mutating the result. Bad idea??` comment for my own hacking notes. But this could re-appear when people write code. 

Perhaps we can take this as a tip to improve the error message or something. I'm sure I'm not the last one to make that kinda mistake in implementation :)

Great work guys!
